src/environments/cd.js

```diff
@@ -0,0 +1,262 @@
+                        if (isLabelEnd(nodes[k], arrowChar)) {
+                            j = k;
+                            break;
+                        }
+                        firstLabel.body.push(nodes[k]);
+                    }
+                    for (let k = j + 1; k < nodes.length; k++) {
+                        if (isLabelEnd(nodes[k], arrowChar)) {
+                            j = k;
+                            break;
+                        }
+                        secondLabel.body.push(nodes[k]);
+                    }
+                } else {
+                    throw new ParseError(`Expected one of "<>AV=|." after @.`);
+                }
+
+                // Now join the arrow to its labels.
+                // This acts in a way similar to a macro expansion.
+                const funcName = cdArrow[arrowChar];
+                let arrow;
+                switch (funcName) {
+                    case "\\\\cdrightarrow":
+                    case "\\\\cdleftarrow":
+                        arrow = parser.callFunction(
+                            funcName, [firstLabel], [secondLabel]
+                        );
+                        break;
+                    case "\\uparrow":
+                    case "\\downarrow": {
+                        const leftLabel = parser.callFunction(
+                            "\\\\cdleft", [firstLabel], []
+                        );
+                        arrow = {
+                            type: "atom",
+                            text: funcName,
+                            mode: "math",
+                            family: "rel",
+                        };
+                        arrow = parser.callFunction("\\Big", [arrow], []);
+                        const rightLabel = parser.callFunction(
+                            "\\\\cdright", [secondLabel], []
+                        );
+                        arrow = {
+                            type: "ordgroup",
+                            mode: "math",
+                            body: [leftLabel, arrow, rightLabel],
+                        };
+                        arrow = parser.callFunction("\\\\cdparent", [arrow], []);
+                        break;
+                    }
+                    case "\\\\cdlongequal":
+                        arrow = parser.callFunction("\\\\cdlongequal", [], []);
+                        break;
+                    case "\\Vert":
+                        arrow = {type: "textord", text: "\\Vert", mode: "math"};
+                        arrow = parser.callFunction("\\Big", [arrow], []);
+                        break;
+                    default:
+                        arrow = {type: "textord", text: " ", mode: "math"};
+                }
+                arrow = {
+                    type: "styling",
+                    body: [arrow],
+                    mode: "math",
+                    style: "display",
+                };
```

Comment:
There's a lot going on in these `for` loops.  This seems like a good candidate to extract into a helper function.

Reply:
Done.